aten/src/ATen/TensorIterator.cpp

```diff
@@ -1358,7 +1358,11 @@ void TensorIteratorBase::build(TensorIteratorConfig& config) {
 
   for (auto& op : operands_) {
     TORCH_INTERNAL_ASSERT(op.tensor->defined());
-    op.data = op.tensor->data_ptr();
+    // Don't access the data_ptr if the tensor is size-0; if there's an offset
+    // into storage that will trigger UB.
+    if (op.tensor->numel() != 0) {
```

Comment:
doesn't look like the right place for the fix, the problem is in https://github.com/pytorch/pytorch/blob/master/c10/core/TensorImpl.h#L1183. I was hoping that tensors without storage don't have storage_offset_, but that's not true for slices.

Reply:
> We could check for is_empty() instead of storage_offset_ there

This seems reasonable to me.  It's in keeping with the semantics of the function  (an index into an empty tensor is still empty, after all), whereas quietly forcing storage_offset_ to 0 seems a bit more "magical".

Reply:
This lgtm, @ezyang if you can take a quick look would be great